go/vt/vttablet/onlineddl/executor.go

```diff
@@ -1185,30 +1454,176 @@ func (e *Executor) dropPTOSCMigrationTriggers(ctx context.Context, onlineDDL *sc
 	return err
 }
 
-// reviewRunningMigrations iterates migrations in 'running' state (there really should just be one that is
-// actually running).
-func (e *Executor) reviewRunningMigrations(ctx context.Context) (countRunnning int, runningNotByThisProcess []string, err error) {
-	e.migrationMutex.Lock()
-	defer e.migrationMutex.Unlock()
+// readVReplStream reads _vt.vreplication entries for given workflow
+func (e *Executor) readVReplStream(ctx context.Context, uuid string, okIfMissing bool) (*VReplStream, error) {
+	query, err := sqlparser.ParseAndBind(sqlReadVReplStream,
+		sqltypes.StringBindVariable(uuid),
+	)
+	if err != nil {
+		return nil, err
+	}
+	r, err := e.execQuery(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	if len(r.Rows) == 0 && okIfMissing {
+		return nil, nil
+	}
+	row := r.Named().Row()
+	if row == nil {
+		return nil, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "Cannot find unique workflow for UUID: %+v", uuid)
+	}
+	s := &VReplStream{
+		id:                   row.AsInt64("id", 0),
+		workflow:             row.AsString("workflow", ""),
+		source:               row.AsString("source", ""),
+		pos:                  row.AsString("pos", ""),
+		timeUpdated:          row.AsInt64("time_updated", 0),
+		transactionTimestamp: row.AsInt64("transaction_timestamp", 0),
+		state:                row.AsString("state", ""),
+		message:              row.AsString("message", ""),
+		bls:                  &binlogdatapb.BinlogSource{},
+	}
+	if err := proto.UnmarshalText(s.source, s.bls); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
 
-	parsed := sqlparser.BuildParsedQuery(sqlSelectRunningMigrations, "_vt", ":strategy")
-	bindVars := map[string]*querypb.BindVariable{
-		"strategy": sqltypes.StringBindVariable(string(schema.DDLStrategyPTOSC)),
+// isVReplMigrationReadyToCutOver sees if the vreplication migration has completed the row copy
+// and is up to date with the binlogs.
+func (e *Executor) isVReplMigrationReadyToCutOver(ctx context.Context, s *VReplStream) (isReady bool, err error) {
```

Comment:
Interesting. Isn't there a heartbeat mechanism for vreplication though?

Reply:
`time_updated` is updated by the heartbeat. However `transaction_timestamp` is only when a new binlog event is seen, not for a heartbeat. This is quite rare since it means there is no writes on the source at all. But it has been reported by some users. Also if a database is taken "offline" for some reason (in the sense that no app is writing to it) and the user starts a workflow this can happen. 

Practically we may not need to address this now.

Reply:
removed evaluation of transaction-timestamp. I want to be on the safer side.

Reply:
`timeUpdated` will always be current because of the heartbeat so the moment the copy state is empty it will say that the migrate workflow is ready for cutover.  Am I missing something?

Reply:
> Am I missing something?

Ermm... Not sure why you're asking? Am _I_ missing something? :sweat_smile:

Reply:
To elaborate a bit, though I'm still unsure what you meant:

> timeUpdated will always be current 

Unless there's some lag, or vreplication is stopped

> so the moment the copy state is empty 

We also validate that `pos` in non-empty. Were you suggesting the migration might cut-over _before starting_ copying rows?

Reply:
Having discussed this: meanwhile checking `transaction_timestamp` is the safer approach. `time_updated` gets update regardless of incoming events.
So the risk of `transaction_timestamp` is in completely stale servers. However, it's enough that _some_ table is updated and the problem goes away. But if the server is completely stale, cut-over will not happen.

To solve that we can either:
- force injection of some dummy change, or
- compare GTID pos (likely not a good idea, because `_vt` tables _are_ getting changed, affecting GTID, but scrubbed in the streamer)
- modify `streamer` to include more information that we can use on target side.